lib/Dialect/SPIRV/SPIRVOps.cpp

```diff
@@ -199,6 +213,136 @@ static void printNoIOOp(Operation *op, OpAsmPrinter *printer) {
+  if (!indices.size()) {
+    return mlir::emitError(baseLoc, "'spv.AccessChain' op expected at least "
+                                    "one index for spv.AccessChain"),
+           nullptr;
+  }
+
+  auto ptrType = type.dyn_cast<spirv::PointerType>();
+  if (!ptrType) {
+    return mlir::emitError(baseLoc, "'spv.AccessChain' op expected a pointer "
+                                    "to composite type, but provided ")
+               << type,
+           nullptr;
+  }
+
+  auto resultType = ptrType.getPointeeType();
+  auto resultStorageClass = ptrType.getStorageClass();
+  int32_t index = 0;
+
+  for (auto indexSSA : indices) {
+    auto cType = resultType.dyn_cast<spirv::CompositeType>();
+    if (!cType) {
+      return mlir::emitError(
+                 baseLoc,
+                 "'spv.AccessChain' op cannot extract from non-composite type ")
+                 << resultType << " with index " << index,
+             nullptr;
+    }
+    index = 0;
+    if (resultType.isa<spirv::StructType>()) {
+      Operation *op = indexSSA->getDefiningOp();
+      if (!op) {
+        return mlir::emitError(baseLoc, "'spv.AccessChain' op index must be an "
+                                        "integer spv.constant to access "
+                                        "element of spv.struct"),
+               nullptr;
+      }
+
+      if (!extractValueFromConstOp(op, index)) {
+        return mlir::emitError(
+                   baseLoc,
+                   "'spv.AccessChain' index must be an integer spv.constant to "
+                   "access element of spv.struct, but provided ")
+                   << op->getName(),
+               nullptr;
+      }
+
+      if (index < 0 || static_cast<uint64_t>(index) >= cType.getNumElements()) {
+        return mlir::emitError(baseLoc, "'spv.AccessChain' op index ")
+                   << index << " out of bounds for " << type,
+               nullptr;
+      }
+    }
+    resultType = cType.getElementType(index);
+  }
+  return spirv::PointerType::get(resultType, resultStorageClass);
+}
+
+static ParseResult parseAccessChainOp(OpAsmParser *parser,
+                                      OperationState *state) {
+  OpAsmParser::OperandType ptrInfo;
+  SmallVector<OpAsmParser::OperandType, 4> indicesInfo;
+  Type type;
+  // TODO(denis0x0D): regarding to the spec an index must be any integer type,
+  // figure out how to use resolveOperand with a range of types and do not
+  // fail on first attempt.
+  Type indicesType = parser->getBuilder().getIntegerType(32);
+
+  if (parser->parseOperand(ptrInfo) ||
+      parser->parseOperandList(indicesInfo, OpAsmParser::Delimiter::Square) ||
+      parser->parseColonType(type) ||
+      parser->resolveOperand(ptrInfo, type, state->operands) ||
+      parser->resolveOperands(indicesInfo, indicesType, state->operands)) {
+    return failure();
+  }
+
+  Location baseLoc = state->operands.front()->getLoc();
+  SmallVector<Value *, 4> indices(state->operands.begin() + 1,
```

Comment:
I think we don't need to materialize a `SmallVector` here? Passing `llvm::makeArrayRef(state->operands).drop_front()` in should do.

Reply:
Oh, thanks, fixed.